### PR TITLE
DM-32519: Removed all Raspberry Pi specific code since setting the GP…

### PR DESCRIPTION
**Pull request by wvreeven**
…IO pins should be handled by the OS. Added unit tests for the FTDI and Raspberry Pi Serial Hat devices.

### CODE REVIEW

**Review by r-owen — APPROVED** [review]
This is a wonderful enhancement. I'm impressed at how much code you were able to get rid of and appreciate the clear documentation on how to set the pins.

I do have a few suggestions I hope you will consider. The most significant is to move the base test class into the library (making it part of the lsst.ess.controller namespace) and add documentation to it. Also I suggest a way to simplify reading data.

Also the existing code to read the serial port seems inefficient. I suggest a few ways to make it more efficient.

File: python/lsst/ts/ess/controller/device/rpi_serial_hat.py [hunk 1]
```diff
@@ -103,96 +58,14 @@ def __init__(
...
-        ----------
-        rpi_pin : `int`
-            The pin number.
-        """
         try:
-            gpio.cleanup(rpi_pin)
-        except RuntimeError:
-            self.log.exception("GPIO pin cleanup error.")
+            self.ser = serial.Serial()
+            self.ser.port = device_id
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/controller/device/rpi_serial_hat.py:63]: Done

File: python/lsst/ts/ess/controller/device/rpi_serial_hat.py [hunk 2]
```diff
@@ -228,11 +101,9 @@ async def readline(self) -> str:
         # get event loop to run blocking tasks
         loop = asyncio.get_event_loop()
 
-        if self._pin_dirn:
-            self._rpi_pin_state(self._pin_dirn, RpiSerialHat.STATE_DIRN_RX)  # type: ignore
-        while not line.endswith(self._sensor.terminator):
-            ch = await loop.run_in_executor(None, self._ser.read, 1)
-            line += ch.decode(self._sensor.charset)
+        while not line.endswith(self.sensor.terminator):
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/controller/device/rpi_serial_hat.py:104]: I tried the `io.TextIOWrapper` method and got a read error so I resigned for the `io.ByteIO` method instead. Thanks for the great suggestion!

File: python/lsst/ts/ess/controller/device/rpi_serial_hat.py [hunk 3]
```diff
@@ -228,11 +101,9 @@ async def readline(self) -> str:
...
+            ch = await loop.run_in_executor(None, self.ser.read, 1)
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/controller/device/rpi_serial_hat.py:105]: Great suggestion, thanks! I made note of it for future reference.

File: python/lsst/ts/ess/controller/device/vcp_ftdi.py [hunk 4]
```diff
@@ -78,11 +78,11 @@ async def basic_open(self) -> None:
         ------
         IOError if virtual communications port fails to open.
         """
-        self._vcp.open()
-        self._vcp.baudrate = 19600
-        if not self._vcp.closed:
+        self.vcp.open()
+        self.vcp.baudrate = 19600
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/controller/device/vcp_ftdi.py:82]: Good point. It is not a constructor argument of `pylibftdi.Device` so I moved the line into the constructor of the VcpFtdi class right after the Device instance gets created.

File: tests/test_rpi_serial_hat.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from lsst.ts.ess import common, controller
+
+logging.basicConfig(
+    format="%(asctime)s:%(levelname)s:%(name)s:%(message)s", level=logging.DEBUG
+)
+
+
+class RpiSerialHatTestCase(BaseRealSensorMockTestCase):
+    @mock.patch("lsst.ts.ess.controller.device.rpi_serial_hat.serial.Serial")
+    async def test_rpi_serial_hat(self, mock_serial: mock.MagicMock) -> None:
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_rpi_serial_hat.py:36]: Thanks for catching that. Done.

File: tests/test_rpi_serial_hat.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        type(device.ser).read = self.read
+        await self.read_task
+        assert self.reply is not None
+        reply_to_check = self.reply[common.Key.TELEMETRY]
+        self.mtt.check_temperature_reply(
+            reply=reply_to_check, name=name, num_channels=self.num_channels
+        )
+
+        self.read_task = asyncio.Future()
+        await self.read_task
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_rpi_serial_hat.py:61]: Done.

File: tests/base_real_sensor_mock_test_case.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BaseRealSensorMockTestCase"]
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/base_real_sensor_mock_test_case.py:22]: I agree and I tried this but I cannot get it to work properly.

File: tests/base_real_sensor_mock_test_case.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import unittest
+
+from lsst.ts.ess import common
+
+logging.basicConfig(
+    format="%(asctime)s:%(levelname)s:%(name)s:%(message)s", level=logging.DEBUG
+)
+
+
+class BaseRealSensorMockTestCase(unittest.IsolatedAsyncioTestCase):
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/base_real_sensor_mock_test_case.py:36]: I added doc strings.

File: tests/base_real_sensor_mock_test_case.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        self, reply: typing.Dict[str, typing.List[typing.Union[str, float]]]
+    ) -> None:
+        self.reply: typing.Optional[
+            typing.Dict[str, typing.List[typing.Union[str, float]]]
+        ] = reply
+        self.read_task.set_result(True)
+        self.sensor_output = None
+
+    def read(self, length: int) -> str:
+        assert length == 1
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/base_real_sensor_mock_test_case.py:59]: The code that calls this method in all cases passes on a value of 1 which is why I added the assert. I suppose I could expand the code to allow for other values as well but I'll do that if the needs arises, which I think will never happen. In any case, I added this explanation to the doc string of `length`.

File: tests/base_real_sensor_mock_test_case.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/base_real_sensor_mock_test_case.py:22]: I'll do this in a future ticket.

File: tests/base_real_sensor_mock_test_case.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/base_real_sensor_mock_test_case.py:39]: Inheriting I'll do in a future ticket.

**Pull request merged**